File: export-results.py
#!/usr/bin/env python3
"""
Stage 03 -> 06 bridge: export-results.py

Reads Playwright's JSON reporter output (test-results/results.json) and
inserts one row per test into fact_testrun, matching schema.sql.

This script does NOT care how the test code was written (Stage 02) or
whether it's a UI test or REST-API test — it just reads whatever Playwright
reports and writes rows. test_case_id is derived from the test title for now;
once Stage 01/02 exist, you'll map real TC-SF-XXXX ids instead (see NOTE below).

Usage:
    python3 export-results.py --results test-results/results.json --db qa.db --ci-pipeline "sf-login-tests.yml" --branch main

Requires: Python 3.8+, no external packages (uses built-in sqlite3 + json).
"""

import argparse
import json
import sqlite3
import sys
import uuid
from datetime import datetime, timezone


def flatten_specs(suites, parent_titles=None):
    """
    Playwright's JSON report nests suites inside suites inside suites.
    This walks the tree and yields (full_title, test_result) pairs,
    flattening the nesting into a single list.
    """
    if parent_titles is None:
        parent_titles = []

    for suite in suites:
        titles = parent_titles + [suite.get("title", "")]

        for spec in suite.get("specs", []):
            spec_title = spec.get("title", "untitled")
            for test in spec.get("tests", []):
                # A test can be retried; results is a list, take the LAST
                # attempt as the final outcome (matches what Playwright shows you).
                results = test.get("results", [])
                if not results:
                    continue
                final_result = results[-1]
                was_retried = len(results) > 1

                yield {
                    "full_title": " > ".join(titles + [spec_title]),
                    "spec_title": spec_title,
                    "status": final_result.get("status", "unknown"),
                    "duration_ms": final_result.get("duration", 0),
                    "was_retried": was_retried,
                }

        # Recurse into nested suites
        if suite.get("suites"):
            yield from flatten_specs(suite["suites"], titles)


def map_status(pw_status: str, was_retried: bool) -> str:
    """
    Maps Playwright's status values to the schema's allowed values:
    Pass / Fail / Flaky / Skipped

    A test that failed at least once but ultimately passed after retry
    is classified as Flaky -- per the rule defined in powerbi-data-model.md:
    "failed, then passed on an immediate retry against the same commit."
    This script can only see the FINAL result per test from the JSON as
    structured above; if you enable Playwright retries in CI, the richer
    flaky detection (checking ALL attempts, not just the last) is the
    next improvement -- flagged in the NOTE at the bottom of this file.
    """
    mapping = {
        "passed": "Pass",
        "failed": "Fail",
        "timedOut": "Fail",
        "skipped": "Skipped",
        "interrupted": "Fail",
    }
    base_status = mapping.get(pw_status, "Fail")
    if was_retried and base_status == "Pass":
        return "Flaky"
    return base_status


def derive_test_case_id(full_title: str) -> str:
    """
    TEMPORARY: derives a stable-ish test_case_id from the test title since
    Stage 01 (spec-to-testcases) doesn't exist yet to assign real TC-SF-XXXX
    ids. Once Stage 01/02 exist, replace this with a real lookup -- e.g. a
    title -> test_case_id mapping table, or tags embedded in the test itself
    (test.describe with a tag like @TC-SF-0001).
    """
    slug = (
        full_title.lower()
        .replace(" > ", "-")
        .replace(" ", "-")
        .replace("'", "")
    )
    slug = "".join(c for c in slug if c.isalnum() or c == "-")[:40]
    return f"TC-SF-AUTO-{slug}"


def ensure_testcase_exists(cur, test_case_id: str, title: str, test_layer: str):
    """
    Inserts a minimal dim_testcase row if one doesn't already exist, so the
    foreign key on fact_testrun doesn't fail. This is a placeholder until
    Stage 01 properly owns dim_testcase population.
    """
    cur.execute(
        "SELECT 1 FROM dim_testcase WHERE test_case_id = ?", (test_case_id,)
    )
    if cur.fetchone() is None:
        cur.execute(
            """
            INSERT INTO dim_testcase
                (test_case_id, title, suite, test_layer, automation_status)
            VALUES (?, ?, ?, ?, 'Automated')
            """,
            (test_case_id, title, "Login Tests", test_layer),
        )


def main():
    parser = argparse.ArgumentParser(description="Export Playwright JSON results into fact_testrun")
    parser.add_argument("--results", required=True, help="Path to Playwright results.json")
    parser.add_argument("--db", required=True, help="Path to SQLite database (schema.sql already applied)")
    parser.add_argument("--ci-pipeline", default="local-run", help="CI workflow name")
    parser.add_argument("--branch", default="main", help="Git branch")
    parser.add_argument("--org-id", default=None, help="Optional dim_org.org_id to attach to these runs")
    args = parser.parse_args()

    try:
        with open(args.results, "r") as f:
            data = json.load(f)
    except FileNotFoundError:
        print(f"ERROR: results file not found at {args.results}", file=sys.stderr)
        sys.exit(1)
    except json.JSONDecodeError as e:
        print(f"ERROR: results file is not valid JSON: {e}", file=sys.stderr)
        sys.exit(1)

    suites = data.get("suites", [])
    if not suites:
        print("WARNING: no suites found in results.json -- nothing to export.")
        sys.exit(0)

    conn = sqlite3.connect(args.db)
    cur = conn.cursor()

    inserted = 0
    status_counts = {"Pass": 0, "Fail": 0, "Flaky": 0, "Skipped": 0}

    for entry in flatten_specs(suites):
        test_case_id = derive_test_case_id(entry["full_title"])
        status = map_status(entry["status"], entry["was_retried"])
        status_counts[status] = status_counts.get(status, 0) + 1

        ensure_testcase_exists(cur, test_case_id, entry["spec_title"], "UI")

        run_id = f"RUN-{uuid.uuid4().hex[:12]}"
        cur.execute(
            """
            INSERT INTO fact_testrun
                (run_id, test_case_id, org_id, run_timestamp, status,
                 duration_ms, ci_pipeline, branch, locator_strategy, http_status_code)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, NULL, NULL)
            """,
            (
                run_id,
                test_case_id,
                args.org_id,
                datetime.now(timezone.utc).isoformat(),
                status,
                entry["duration_ms"],
                args.ci_pipeline,
                args.branch,
            ),
        )
        inserted += 1

    conn.commit()
    conn.close()

    print(f"Exported {inserted} test run(s) into {args.db}")
    print(f"Status breakdown: {status_counts}")


if __name__ == "__main__":
    main()

File: reporting/export-results.ts
/**
 * export-results.ts
 *
 * TypeScript rewrite of export-results.py. Reads Playwright's JSON reporter
 * output (test-results/results.json) and inserts one row per test into
 * fact_testrun, matching schema.sql. Same logic as the Python version --
 * just no Python required.
 *
 * Uses sql.js (WASM SQLite, no native compilation) -- same library as
 * export-to-csv.ts, so this whole pipeline now shares one dependency.
 *
 * IMPORTANT: sql.js loads the entire database into memory and only writes
 * changes back to disk when you explicitly export and save the buffer.
 * This script handles that automatically -- it loads qa.db, makes changes,
 * then writes the updated buffer back to the same file. Don't run two
 * instances of this script against the same qa.db at the same time, since
 * the second one would overwrite the first's changes (same risk applies
 * to the Python version too, just worth knowing either way).
 *
 * Usage:
 *   npm run build
 *   node export-results.js --results test-results/results.json --db qa.db --ci-pipeline "sf-login-tests" --branch main --org-id SBX-UAT-01
 *
 * Requires: npm install sql.js
 *           npm install -D typescript @types/node @types/sql.js
 */

import initSqlJs, { Database } from "sql.js";
import * as fs from "fs";
import * as crypto from "crypto";

interface CliArgs {
  results: string;
  db: string;
  ciPipeline: string;
  branch: string;
  orgId: string | null;
}

interface PlaywrightResult {
  status: string;
  duration: number;
}

interface PlaywrightTest {
  results: PlaywrightResult[];
}

interface PlaywrightSpec {
  title: string;
  tests: PlaywrightTest[];
}

interface PlaywrightSuite {
  title: string;
  file?: string;
  specs?: PlaywrightSpec[];
  suites?: PlaywrightSuite[];
}

interface PlaywrightReport {
  suites: PlaywrightSuite[];
}

interface FlatTestEntry {
  fullTitle: string;
  specTitle: string;
  status: string;
  durationMs: number;
  wasRetried: boolean;
  filePath: string;
  topLevelSuiteTitle: string;
}

function parseArgs(argv: string[]): CliArgs {
  const args: Record<string, string> = {};
  for (let i = 0; i < argv.length; i++) {
    const key = argv[i];
    if (key.startsWith("--")) {
      args[key.slice(2)] = argv[++i];
    }
  }
  if (!args.results) {
    console.error("ERROR: --results <path> is required");
    process.exit(1);
  }
  if (!args.db) {
    console.error("ERROR: --db <path> is required");
    process.exit(1);
  }
  return {
    results: args.results,
    db: args.db,
    ciPipeline: args["ci-pipeline"] ?? "local-run",
    branch: args.branch ?? "main",
    orgId: args["org-id"] ?? null,
  };
}

/**
 * Walks Playwright's nested suite structure recursively and yields a flat
 * list of test entries. Mirrors flatten_specs() from the Python version,
 * with one addition: tracks the top-level suite's `file` path and title
 * as it recurses, so callers can tell which file/suite each test actually
 * came from -- needed to correctly distinguish UI vs API tests instead
 * of hardcoding one value for everything.
 */
function* flattenSpecs(
  suites: PlaywrightSuite[],
  parentTitles: string[] = [],
  topLevelFile: string = "",
  topLevelSuiteTitle: string = ""
): Generator<FlatTestEntry> {
  for (const suite of suites) {
    const titles = [...parentTitles, suite.title ?? ""];
    // The file path only appears on the OUTERMOST suite in Playwright's
    // JSON (e.g. the suite representing the whole .spec.ts file). Nested
    // suites (describe blocks) don't repeat it, so once we're past the
    // top level we keep carrying forward whatever was captured there.
    const currentFile = parentTitles.length === 0 ? (suite.file ?? "") : topLevelFile;
    const currentTopLevelTitle = parentTitles.length === 0 ? (suite.title ?? "") : topLevelSuiteTitle;

    for (const spec of suite.specs ?? []) {
      const specTitle = spec.title ?? "untitled";
      for (const test of spec.tests ?? []) {
        const results = test.results ?? [];
        if (results.length === 0) continue;
        const finalResult = results[results.length - 1];
        const wasRetried = results.length > 1;

        yield {
          fullTitle: [...titles, specTitle].join(" > "),
          specTitle,
          status: finalResult.status ?? "unknown",
          durationMs: finalResult.duration ?? 0,
          wasRetried,
          filePath: currentFile,
          topLevelSuiteTitle: currentTopLevelTitle,
        };
      }
    }

    if (suite.suites) {
      yield* flattenSpecs(suite.suites, titles, currentFile, currentTopLevelTitle);
    }
  }
}

/**
 * Determines test_layer (UI vs REST-API) from the test file's path.
 * Anything under a folder literally named "api" (e.g. tests/api/lead.api.spec.ts,
 * or Windows-style tests\api\lead.api.spec.ts) is REST-API; everything
 * else defaults to UI. This is a simple, explicit convention -- if you
 * organize test files differently, update this function to match.
 */
function deriveTestLayer(filePath: string): string {
  const normalized = filePath.replace(/\\/g, "/").toLowerCase();
  if (normalized.includes("/api/") || normalized.startsWith("api/")) {
    return "REST-API";
  }
  return "UI";
}

/**
 * Maps Playwright's status values to the schema's allowed values:
 * Pass / Fail / Flaky / Skipped
 *
 * A test that ultimately passed but only after a retry is classified as
 * Flaky, matching the rule in powerbi-data-model.md: "failed, then passed
 * on an immediate retry against the same commit/branch."
 */
function mapStatus(pwStatus: string, wasRetried: boolean): string {
  const mapping: Record<string, string> = {
    passed: "Pass",
    failed: "Fail",
    timedOut: "Fail",
    skipped: "Skipped",
    interrupted: "Fail",
  };
  const baseStatus = mapping[pwStatus] ?? "Fail";
  if (wasRetried && baseStatus === "Pass") {
    return "Flaky";
  }
  return baseStatus;
}

/**
 * TEMPORARY: derives a stable-ish test_case_id from the test title since
 * Stage 01 (spec-to-testcases) doesn't exist yet to assign real TC-SF-XXXX
 * ids. Once Stage 01/02 exist, replace this with a real lookup.
 */
function deriveTestCaseId(fullTitle: string): string {
  let slug = fullTitle
    .toLowerCase()
    .replace(/ > /g, "-")
    .replace(/ /g, "-")
    .replace(/'/g, "");
  slug = slug
    .split("")
    .filter((c) => /[a-z0-9-]/.test(c))
    .join("")
    .slice(0, 40);
  return `TC-SF-AUTO-${slug}`;
}

/**
 * Inserts a minimal dim_testcase row if one doesn't already exist, so the
 * foreign key on fact_testrun doesn't fail. Placeholder until Stage 01
 * properly owns dim_testcase population.
 */
function ensureTestcaseExists(
  db: Database,
  testCaseId: string,
  title: string,
  suiteName: string,
  testLayer: string
): void {
  const checkStmt = db.prepare("SELECT 1 FROM dim_testcase WHERE test_case_id = :id");
  checkStmt.bind({ ":id": testCaseId });
  const exists = checkStmt.step();
  checkStmt.free();

  if (!exists) {
    const insertStmt = db.prepare(`
      INSERT INTO dim_testcase
        (test_case_id, title, suite, test_layer, automation_status)
      VALUES (:id, :title, :suite, :layer, 'Automated')
    `);
    insertStmt.run({
      ":id": testCaseId,
      ":title": title,
      ":suite": suiteName,
      ":layer": testLayer,
    });
    insertStmt.free();
  }
}

function generateRunId(): string {
  return `RUN-${crypto.randomBytes(6).toString("hex")}`;
}

async function main(): Promise<void> {
  const args = parseArgs(process.argv.slice(2));

  let reportData: PlaywrightReport;
  try {
    const raw = fs.readFileSync(args.results, "utf-8");
    reportData = JSON.parse(raw);
  } catch (err) {
    console.error(`ERROR: could not read or parse ${args.results}: ${(err as Error).message}`);
    process.exit(1);
    return;
  }

  const suites = reportData.suites ?? [];
  if (suites.length === 0) {
    console.log("WARNING: no suites found in results.json -- nothing to export.");
    return;
  }

  if (!fs.existsSync(args.db)) {
    console.error(`ERROR: database file not found at ${args.db}`);
    process.exit(1);
    return;
  }

  const SQL = await initSqlJs();
  const fileBuffer = fs.readFileSync(args.db);
  const db = new SQL.Database(fileBuffer);

  let inserted = 0;
  const statusCounts: Record<string, number> = { Pass: 0, Fail: 0, Flaky: 0, Skipped: 0 };

  const insertRunStmt = db.prepare(`
    INSERT INTO fact_testrun
      (run_id, test_case_id, org_id, run_timestamp, status,
       duration_ms, ci_pipeline, branch, locator_strategy, http_status_code)
    VALUES (:run_id, :test_case_id, :org_id, :run_timestamp, :status,
            :duration_ms, :ci_pipeline, :branch, NULL, NULL)
  `);

  for (const entry of flattenSpecs(suites)) {
    const testCaseId = deriveTestCaseId(entry.fullTitle);
    const status = mapStatus(entry.status, entry.wasRetried);
    statusCounts[status] = (statusCounts[status] ?? 0) + 1;

    ensureTestcaseExists(
      db,
      testCaseId,
      entry.specTitle,
      entry.topLevelSuiteTitle,
      deriveTestLayer(entry.filePath)
    );

    const runId = generateRunId();
    insertRunStmt.run({
      ":run_id": runId,
      ":test_case_id": testCaseId,
      ":org_id": args.orgId,
      ":run_timestamp": new Date().toISOString(),
      ":status": status,
      ":duration_ms": entry.durationMs,
      ":ci_pipeline": args.ciPipeline,
      ":branch": args.branch,
    });
    inserted++;
  }

  insertRunStmt.free();

  // sql.js keeps everything in memory -- this writes the updated database
  // back to disk, overwriting the original file with the new rows included.
  const updatedBuffer = db.export();
  fs.writeFileSync(args.db, Buffer.from(updatedBuffer));
  db.close();

  console.log(`Exported ${inserted} test run(s) into ${args.db}`);
  console.log(`Status breakdown: ${JSON.stringify(statusCounts)}`);
}

main().catch((err) => {
  console.error("FATAL:", err);
  process.exit(1);
});
